client/src/partials/enterprise/enterprise.js

```diff
@@ -1,104 +1,109 @@
-      $scope.globalEnterprise = enterprise;
-      validate.process(dependencies)
-      .then(initialize);
-    });
-
-    function initialize (models) {
-      angular.extend($scope, models);
-      $scope.newAccount = {};
-    }
-
-    $scope.formatLocation = function formatLocation (l) {
-      return [l.name, l.sector_name, l.province_name, l.country_name].join(' -- ');
-    };
-
-    $scope.newEnterprise = function () {
-      $scope.add = {};
-      $scope.action = 'new';
-    };
-
-    $scope.editEnterprise = function (enterprise) {
-      $scope.edit = angular.copy(enterprise);
-      $scope.editingEnterprise = enterprise;
-      $scope.action = 'edit';
-    };
-
-    $scope.saveEdit = function () {
-      var data = connect.clean($scope.edit);
-
-      connect.put('enterprise', [data], ['id'])
-      .then(function () {
-        $scope.enterprise.put(data);
-        $scope.action = '';
-
-        // we should reset the global enterprise variable
-        // if we have updated the global enterprise data
-        if (data.id === $scope.globalEnterprise.id) {
-          appstate.set('enterprise', data);
-        }
+.controller('EnterpriseController', EnterpriseController);
 
-      });
-    };
+EnterpriseController.$inject = [
+  'EnterpriseService', 'CurrencyService', 'FormStateFactory'
+];
 
-    $scope.resetEdit = function () {
-      $scope.edit = angular.copy($scope.editingEnterprise);
-    };
+function EnterpriseController(Enterprises, Currencies, StateFactory) {
+  var vm = this;
 
-    $scope.saveNew = function () {
-      var data = connect.clean($scope.add);
+  vm.enterprises = [];
+  vm.state = new StateFactory();
+  vm.view = 'default';
 
-      connect.post('enterprise', [data])
-      .then(function (res) {
-        data.id = res.data.insertId;
-        $scope.enterprise.post(data);
-        $scope.action = '';
-      });
-    };
+  // bind methods
+  vm.create = create;
+  vm.update = update;
+  vm.cancel = cancel;
+  vm.submit = submit;
+  vm.formatLocation = formatLocation;
+  vm.getElementCurrency = getElementCurrency;
+
+  /* ------------------------------------------------------------------------ */
+
+  function handler(error) {
+    console.error(error);
+    vm.state.error();
+  }
+
+  // fired on startup
+  function startup() {
+    // load Enterprises
+    Enterprises.read().then(function (data) {
+      vm.enterprises = data;
+    }).catch(handler);
 
-    $scope.resetNew = function () {
-      $scope.add = {};
-    };
+    Enterprises.readLocations().then(function (data) {
+      vm.locations = data;
+    }).catch(handler);
 
-    $scope.print = function () {
-      $window.print();
-    };
+    Currencies.read().then(function (data) {
+      vm.currencies = data;
+    }).catch(handler);
+
+    vm.state.reset();
+  }
+
+  function cancel() {
+    vm.state.reset();
+    vm.view = 'default';
+  }
+
+  function create() {
+    vm.view = 'create';
+    vm.enterprise = {};
+  }
 
+  function formatLocation(l) {
```

Comment:
This function will be run on every view update.  This can happen several times a second.  It is better to format location names once, when the data is downloaded, and bind it to a property.  For example:
```
LocationService.read().then(function (locations) {
  locations.forEach(function (item) {
    item.fmtLabel = [l.name, l.sector_name, l.province_name, l.country_name].join(' -- ');
  });
  vm.locations = locations;
});
```
... and then use it in the view as ...
```html
<tr ng-repeat="location in locations">
  <td>{{ locaton.fmtLabel }}</td>
</tr>
```
This way, the formatting will only happen once.  Angular will use a property for all future lookups.